torch/optim/adam.py

```diff
@@ -429,13 +437,35 @@ def _single_tensor_adam(
         # Decay the first and second moment running average coefficient
         exp_avg.lerp_(grad, 1 - device_beta1)
 
-        exp_avg_sq.mul_(beta2).addcmul_(grad, grad.conj(), value=1 - beta2)
+        # Nested if is necessary to bypass jitscript rules
+        if differentiable and isinstance(beta2, Tensor):
+            if beta2.requires_grad:
+                exp_avg_sq.mul_(beta2).addcmul_(grad, grad.conj().mul(1 - beta2))
```

Comment:
Nested if is fine.

for slightly faster compute in the requires grad path, let’s separate out grad squared into its own variable and use lerp :) that should use 2 ops instead of 3 and requires fewer intermediates.

Reply:
Ok that sounds good with me! I'll send that in soon :)

Reply:
And the lerp trick is really cool!

Reply:
A minor downside is that maybe it would require a temp copy of the grad-sized tensor to store the grad square...

Reply:
Yea, tho the grad.conj().mul(1 - beta2) is already an introduced Tensor of size grad here...I guess it's not that much fewer intermediates if what we're saving is the 1-beta2 lol. The fusion is still worth it.